scalafix-core/shared/src/main/scala/scalafix/internal/rule/SingleAbstractMethod.scala

```diff
@@ -0,0 +1,273 @@
+package scalafix.internal.rule
+
+import scala.meta._
+import scala.meta.Token._
+
+import scala.meta.internal.tokens.TokenInfo
+
+import scalafix.Patch
+import scalafix.SemanticdbIndex
+import scalafix.rule.RuleCtx
+import scalafix.rule.SemanticRule
+
+import scala.collection.mutable
+
+import scalafix.syntax._
+
+// SAM conversion: http://www.scala-lang.org/files/archive/spec/2.12/06-expressions.html#sam-conversion
+case class SingleAbstractMethod(index: SemanticdbIndex)
+    extends SemanticRule(index, "SingleAbstractMethod") {
+  override def description: String = ???
+  override def fix(ctx: RuleCtx): Patch = {
+    val visited = mutable.Set.empty[Tree]
+    def isSam(tree: Tree): Boolean =
+      tree match {
+        case Term.NewAnonymous(
+            Template(
+              _,
+              List(Init(clazz, _, _)),
+              _,
+              List(Defn.Def(_, method, _, paramss, _, _))
+            )
+            )
+            if !visited.contains(tree) && (paramss.size == 1 || paramss.size == 0) =>
+          visited += tree
```

Comment:
I think this would be more readable with a 

```scala
object SAMCandidate { def unapply(tree: Term.NewAnonymous): Option[(A, B, C)]


var patch = Patch.empty
new Traverser { 
  def apply(tree: Tree) = tree match {
    case SAMCandidate(a, b, c) =>
      patch += samPatch(a, b, c)
      super.apply(c)
    case _ => super.apply(t)
  }
}
patch
```
instead of using `visited: mutable.Set` and `.collect`